akka-actor-typed/src/main/scala/akka/actor/typed/internal/delivery/ConsumerController.scala

```diff
@@ -0,0 +1,414 @@
+ *
+ * The `ConsumerController` sends [[ProducerController.Internal.Request]] to the `ProducerController`
+ * to specify it's ready to receive up to the requested sequence number.
+ *
+ * The `ConsumerController` sends the first `Request` when it receives the first `SequencedMessage`
+ * and has received the `Start` message from the consumer.
+ *
+ * It sends new `Request` when half of the requested window is remaining, but it also retries
+ * the `Request` if no messages are received because that could be caused by lost messages.
+ *
+ * Apart from the first message the producer will not send more messages than requested.
+ *
+ * Received messages are wrapped in [[ConsumerController.Delivery]] when sent to the consumer,
+ * which is supposed to reply with [[ConsumerController.Confirmed]] when it has processed the message.
+ * Next message is not delivered until the previous is confirmed.
+ * More messages from the producer that arrive while waiting for the confirmation are stashed by
+ * the `ConsumerController` and delivered when previous message was confirmed.
+ *
+ * In other words, the "request" protocol to the application producer and consumer is one-by-one, but
+ * between the `ProducerController` and `ConsumerController` it's window of messages in flight.
+ *
+ * The consumer and the `ConsumerController` are supposed to be local so that these messages are fast and not lost.
+ *
+ * If the `ConsumerController` receives a message with unexpected sequence number (not previous + 1)
+ * it sends [[ProducerController.Internal.Resend]] to the `ProducerController` and will ignore all messages until
+ * the expected sequence number arrives.
+ */
+object ConsumerController {
+
+  sealed trait InternalCommand
+  sealed trait Command[+A] extends InternalCommand
+  final case class SequencedMessage[A](producerId: String, seqNr: Long, msg: A, first: Boolean, ack: Boolean)(
+      /** INTERNAL API */
+      @InternalApi private[akka] val producer: ActorRef[ProducerController.InternalCommand])
+      extends Command[A]
+  private final case object Retry extends InternalCommand
+
+  final case class Delivery[A](producerId: String, seqNr: Long, msg: A, confirmTo: ActorRef[Confirmed])
+  final case class Start[A](deliverTo: ActorRef[Delivery[A]]) extends Command[A]
+  final case class Confirmed(seqNr: Long) extends InternalCommand
+
+  final case class RegisterToProducerController[A](producerController: ActorRef[ProducerController.Command[A]])
+      extends Command[A]
+
+  private final case class ConsumerTerminated(consumer: ActorRef[_]) extends InternalCommand
+
+  private final case class State[A](
+      producer: ActorRef[ProducerController.InternalCommand],
+      consumer: ActorRef[ConsumerController.Delivery[A]],
+      receivedSeqNr: Long,
+      confirmedSeqNr: Long,
+      requestedSeqNr: Long)
+
+  private val RequestWindow = 20 // FIXME should be a param, ofc
+
+  def apply[A](resendLost: Boolean): Behavior[Command[A]] = {
```

Comment:
In what scenario would you not want to resend lost, doesn't that take make it unreliable rather than reliable?

Reply:
yes, then it's not reliable but still has flow control, which is useful in itself

if we think it adds too much complexity to support both we could remove that part

Reply:
the main performance benefit of the flow-control-only mode is that the PC doesn't have to buffer pending messages

Reply:
Ah, now I remember talking about the flow control as well previously. Thx for clarifying.

Reply:
there are 3 levels
* flow control (not reliable)
* in-memory reliability
* durable reliability